### PR TITLE
[ENH] Allow Decoder to work with surface objects

### PR DESCRIPTION
**Pull request by ymzayek**
- Part of #4027

Changes proposed in this pull request:

- [x] Add tests for decoder methods with surface object
- [x] Adapt screening percentile function
- [ ] Adapt FREM and ReNA clustering --> moved to #4205
- [x] raise NotImplemented (or similar) errors if users want to use SurfaceImage with
  - [x] space net
  - [ ] FREM (for now)
- Left untouched: searchlight

### CODE REVIEW

**Comment by bthirion** [comment]
Adapting ReNA (and Ward)  to the surface is  technically  easy. We should definitely do that.

**Comment by ymzayek** [comment]
> Adapting ReNA (and Ward) to the surface is technically easy. We should definitely do that.

Well yes and no. I think writing a new image check function that wraps check_niimg and handles surface image separately (we don't want to call `load_niimg`) might make this and a lot of other cases easier.

**Comment by bthirion** [comment]
Ye, but behind that, the algorithm will be a bit different, given that the connectivity structure is then defined by the mesh and no longer by the voxel grid.

**Comment by ymzayek** [comment]
What would be the equivalent calculation of edges and weights for surface data? `_make_edges_and_weights` is a blocker for FREM using ReNA and it depends on several functions that need volumetric data so maybe it would be preferable to write a `_make_edges_and_weights_surface` function that is called from `_weighted_connectivity_graph` when the input is a surface object

**Comment by ymzayek** [comment]
The comment from https://github.com/nilearn/nilearn/pull/4227#pullrequestreview-1828226305 is meant to be here

> We probably also have to adapt Decoder heuristics with the percentile selection ? What about clustering ?

File: nilearn/masking.py [hunk 1]
```diff
@@ -991,7 +991,7 @@ def unmask_from_to_3d_array(w, mask):
     w : :class:`numpy.ndarray`, shape (n_features,)
       The image to be unmasked.
 
-    mask : :class:`numpy.ndarray`, shape (nx, ny, nz)
+    mask : :class:`numpy.ndarray`
```
Comment (ymzayek) [inline nilearn/masking.py:994]: there is nothing in this function that requires that shape

File: nilearn/regions/rena_clustering.py [hunk 2]
```diff
@@ -171,7 +295,12 @@ def _weighted_connectivity_graph(X, mask_img):
     """
     n_features = X.shape[1]
 
-    edges, weight = _make_edges_and_weights(X, mask_img)
+    if isinstance(mask_img, SurfaceImage):
+        edges, weight = _make_edges_and_weights_surface(X, mask_img)
+    else:
+        edges, weight = _make_edges_and_weights(X, mask_img)
+
+    # TODO : deal with dict results if surface analysis
```
Comment (ymzayek) [inline nilearn/regions/rena_clustering.py:303]: Right now failure is here because edges and weights are returned in a dictionary with keys representing the surface image object mesh and data keys for the different hemispheres

**Comment by ymzayek** [comment]
TODO:
- [ ] Finish rena clustering adaptation
- [ ] validate screening percentile and clustering for surface

File: nilearn/decoding/space_net.py [hunk 3]
```diff
@@ -840,6 +840,11 @@ def fit(self, X, y):
         self : `SpaceNet` object
             Model selection is via cross-validation with bagging.
         """
+        if isinstance(X, SurfaceImage) or isinstance(self.mask, SurfaceMasker):
+            raise NotImplementedError(
+                "Running space net on surface objects is not yet supported."
```
Comment (bthirion) [inline nilearn/decoding/space_net.py:845]: We have no intention to support it.

File: nilearn/regions/rena_clustering.py [hunk 4]
```diff
@@ -150,6 +151,129 @@ def _make_edges_and_weights(X, mask_img):
...
+
+def _make_edges_surface(faces, mask):
+    """Create the edges set: Returns a list of edges for a surface mesh.
+
+    Parameters
+    ----------
+    faces : ndarray
+        The vertex indices corresponding the mesh triangles.
+
+    mask : boolean
```
Comment (bthirion) [inline nilearn/regions/rena_clustering.py:196]: I don't see how to make sense of mask in this function, given that the edge ordering is unclear at that point ?

Comment (bthirion) [reply nilearn/regions/rena_clustering.py:196]: maks is on vertices, right ? Then it makes sense.  But the dosctring has to be improved.

**Review by bthirion — COMMENTED** [review]
Thx  ! Thi is great

File: nilearn/regions/rena_clustering.py (hunk 2 above)
Comment (bthirion) [reply nilearn/regions/rena_clustering.py:303]: We could great a unique graph that would stack the data from the two hemispheres. This graph would have 2 connected components. This would have the advantage to make the use on surface transparent for other functions.

File: nilearn/regions/rena_clustering.py [hunk 5]
```diff
@@ -171,11 +291,20 @@ def _weighted_connectivity_graph(X, mask_img):
     """
     n_features = X.shape[1]
 
-    edges, weight = _make_edges_and_weights(X, mask_img)
+    if isinstance(mask_img, SurfaceImage):
```
Comment (ymzayek) [inline nilearn/regions/rena_clustering.py:294]: I did not finish this section under the conditional for surface. It probably needs to be updated according to this comment https://github.com/nilearn/nilearn/pull/4205#discussion_r1461818983

And then the whole pipeline needs to be validated

**Review by bthirion — COMMENTED** [review]
There remain a few comments to address, but it looks good overall.

**Comment by Remi-Gau** [comment]
> Examples to adapt:
> 
>     * [ ]  examples/02_decoding/plot_haxby_searchlight_surface.py

@man-shu 

I was thinking that if you want to get familiar with the new surface object API, trying to adapt this example to use it may be a good way to get you started.

what do you think?

**Comment by man-shu** [comment]
> > Examples to adapt:
> > ```
> > * [ ]  examples/02_decoding/plot_haxby_searchlight_surface.py
> > ```
> 
> @man-shu
> 
> I was thinking that if you want to get familiar with the new surface object API, trying to adapt this example to use it may be a good way to get you started.
> 
> what do you think?

Yes! Sounds great. Thanks, I will look at it.

**Comment by Remi-Gau** [comment]
FYI I think it may be necessary to create a "dataset fetcher" for the haxby dataset that returns surface object like was done for a couple of other fetchers for NKI and destrieux.

See for example:

https://github.com/nilearn/nilearn/blob/88be15f61bb9d794d643dbd3e98239239e3ee13d/nilearn/experimental/surface/_datasets.py#L91

File: nilearn/decoding/tests/test_decoder.py [hunk 6]
```diff
@@ -1086,3 +1076,108 @@ def test_decoder_decision_function_raises_value_error(
...
+    """Test classification predict and scoring for surface image."""
+    X, y = _make_surface_class_data()
+    model = Decoder(mask=SurfaceMasker())
+    model.fit(X, y)
+    y_pred = model.predict(X)
+
+    assert model.scoring == "roc_auc"
+
+    model.score(X, y)
+    accuracy_score(y, y_pred)
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1162]:
```suggestion
    acc = accuracy_score(y, y_pred)
    assert 0.3 < acc < 0.7
```

File: nilearn/decoding/tests/test_decoder.py [hunk 7]
```diff
@@ -1086,3 +1076,108 @@ def test_decoder_decision_function_raises_value_error(
...
+    """Test regression predict and scoring for surface image."""
+    X, y = _make_surface_reg_data()
+    model = DecoderRegressor(mask=SurfaceMasker())
+    model.fit(X, y)
+    y_pred = model.predict(X)
+
+    assert model.scoring == "r2"
+
+    model.score(X, y)
+    r2_score(y, y_pred)
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1175]:
```suggestion
    r2 = r2_score(y, y_pred)
    assert r2 <= 0
```

**Comment by bthirion** [comment]
Is this ready for a review ?

**Comment by Remi-Gau** [comment]
> Is this ready for a review ?

not quite

**Comment by Remi-Gau** [comment]
will move the clustering things to a different PR

**Comment by Remi-Gau** [comment]
blocked by #4577

**Comment by Remi-Gau** [comment]
> Adapting ReNA (and Ward) to the surface is technically easy. We should definitely do that.

Will extract this part in a separate PR to make things easier to code and review

**Comment by Remi-Gau** [comment]
Not mentioned here so far: doing searchlight with SurfaceImages. I would say that this is beyond the scope of this PR.

**Comment by Remi-Gau** [comment]
> Not mentioned here so far: doing searchlight with SurfaceImages. I would say that this is beyond the scope of this PR.

**Comment by Remi-Gau** [comment]
OK this should be good for one final round of reviews.

**Review by bthirion — APPROVED** [review]
LGTM, thx.

File: nilearn/_utils/param_validation.py [hunk 8]
```diff
@@ -221,6 +223,11 @@ def check_feature_screening(
             f" [0, 100], got {screening_percentile:g}"
         )
     else:
+        if isinstance(mask_img, SurfaceImage):
+            return SelectPercentile(
+                f_test, percentile=int(screening_percentile)
+            )
+
```
Comment (man-shu) [inline nilearn/_utils/param_validation.py:230]: I am wondering what would happen if the `mask_img` here already has fewer voxels/vertices than the `screening_percentile`.

With the nii image case, if this is the case, `adjust_screening_percentile` ensures that all the voxels are selected as features but we skip that step here...

Comment (bthirion) [reply nilearn/_utils/param_validation.py:230]: This means that we need a specific unit test, and possibly a fix.

Comment (man-shu) [reply nilearn/_utils/param_validation.py:230]: Ok so as of now, it picks `screening percentile` of the `mask_img`:

The `mask_img`:
output

The selected vertices after screening:
selected

File: nilearn/decoding/tests/test_decoder.py [hunk 9]
```diff
@@ -1089,3 +1079,156 @@ def test_decoder_decision_function_raises_value_error(
...
+
+@pytest.mark.parametrize("decoder", [_BaseDecoder, Decoder, DecoderRegressor])
+def test_decoder_error_incompatible_surface_data_and_volume_mask(
+    _make_surface_class_data, decoder, tiny_binary_classification_data
+):
+    """Test error when fiting for surface data with a volume mask."""
+    data_surface, y = _make_surface_class_data()
+    _, _, mask = tiny_binary_classification_data
+    model = decoder(mask=mask)
+    model.fit(data_surface, y)
```
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_decoder.py:1193]: - Those are currently failing and I would prefer to merge https://github.com/nilearn/nilearn/pull/4603 first so its content can be reused to validate the compatibility between data and maskers.

File: nilearn/_utils/param_validation.py [hunk 10]
```diff
@@ -171,15 +211,19 @@ def adjust_screening_percentile(screening_percentile, mask_img, verbose=0):
...
     return screening_percentile
 
 
 def check_feature_screening(
-    screening_percentile, mask_img, is_classification, verbose=0
+    screening_percentile,
+    mask_img,
+    is_classification,
+    verbose=0,
+    mesh_n_vertices=None,
```
[the remark this answers is not in the document]
Comment (man-shu) [reply nilearn/_utils/param_validation.py:226]: So when `mask_img` has lesser vertices than the given `screening_percentile` value, then the screening considers all vertices of the mask.

[the remark this answers is not in the document]
Comment (man-shu) [reply nilearn/_utils/param_validation.py:226]: Now selected features, for comparison with the previous discussion:
output

**Comment by Remi-Gau** [comment]
The doc build failure is unrelated: also happens on main and I got a PR to fix it.

File: nilearn/_utils/tests/test_param_validation.py [hunk 11]
```diff
@@ -101,3 +105,45 @@ def test_feature_screening(affine_eye):
...
+    )
+
+    if screening_percentile == 100:
+        adjusted = 100
+    if mask_to_mesh_ratio < screening_percentile:
+        adjusted = 100
+    elif mask_to_mesh_ratio > screening_percentile:
+        adjusted = screening_percentile * (mesh_n_vertices / mask_n_vertices)
+    elif mask_to_mesh_ratio == screening_percentile:
+        adjusted = screening_percentile
```
Comment (man-shu) [inline nilearn/_utils/tests/test_param_validation.py:149]: I wanted to test the screening percentile adjustment somewhat like this. However, I am struggling to create an appropriate sample mask and image.

I saw that there are some fixtures defined in `nilearn/experimental/conftest.py` but I couldn't call them here (even after importing them in `nilearn/conftest.py`).

Any suggestions @Remi-Gau ?

Comment (man-shu) [reply nilearn/_utils/tests/test_param_validation.py]: Updated the code with comments:
```python
def test_screening_adjustment(affine_eye, screening_percentile):

    # get a mesh
    fsaverage = datasets.fetch_surf_fsaverage("fsaverage5")
    # create sample mask image
    mask_img_data = np.zeros((182, 218, 182))
    mask_img_data[30:-30, 30:-30, 30:-30] = 1
    mask_img = Nifti1Image(mask_img_data, affine=affine_eye)
    mask_surf = vol_to_surf(
        img=mask_img,
        surf_mesh=fsaverage["pial_left"],
        inner_mesh=fsaverage["white_left"],
    )
    # create sample image
    img_data = np.random.randn(182, 218, 182)
    img = Nifti1Image(img_data, affine=affine_eye)
    img_surf = vol_to_surf(
        img=img,
        surf_mesh=fsaverage["pial_left"],
        inner_mesh=fsaverage["white_left"],
    )
    # get number of vertices in mask and mesh
    mask_n_vertices = get_mask_volume(mask_surf)
    mesh_n_vertices = fsaverage.mesh.n_vertices
    # get mask to mesh ratio
    mask_to_mesh_ratio = (mask_n_vertices / mesh_n_vertices) * 100
    # adjust screening percentile
    adjusted = adjust_screening_percentile(
        screening_percentile, mask_to_mesh_ratio, mesh_n_vertices
    )
    if screening_percentile == 100:
        assert adjusted == 100
    # if mask is smaller than given percentile, full mask is used
    if mask_to_mesh_ratio < screening_percentile:
        assert adjusted == 100
    # if mask is larger than given percentile, the percentile is adjusted to
    # the ratio of mesh to mask
    elif mask_to_mesh_ratio > screening_percentile:
        assert adjusted == screening_percentile * (
            mesh_n_vertices / mask_n_vertices
        )
    # if mask is equal to given percentile, the adjusted percentile is the
    # same as the given percentile
    elif mask_to_mesh_ratio == screening_percentile:
        assert adjusted == screening_percentile
```

Comment (man-shu) [reply nilearn/_utils/tests/test_param_validation.py]: This could be deferred to a separate issue for later. Let me know what you think!

Comment (Remi-Gau) [reply nilearn/_utils/tests/test_param_validation.py]: Yeah I am tempted to put this in a separate issue / PR.
Would rather have smaller PRs.

Comment (man-shu) [reply nilearn/_utils/tests/test_param_validation.py]: Ok I will open a separate issue then

**Review by man-shu — APPROVED** [review]
LGTM!

File: nilearn/decoding/decoder.py [hunk 12]
```diff
@@ -935,9 +949,17 @@ class would be predicted.
 
     def _apply_mask(self, X):
         masker_type = "nii"
-        if isinstance(self.mask, SurfaceMasker):
+        # all elements of X should of the similar type by now
```
Comment (bthirion) [inline nilearn/decoding/decoder.py:952]:
```suggestion
        # all elements of X should be of the similar type by now
```

**Review by bthirion — COMMENTED** [review]
Almost done, thx !

**Comment by Remi-Gau** [comment]
the doc build is stalling on this PR and a few others.

as it was failing on main recently I'd like to know if this is now fixed
